src/mlpack/methods/decision_tree/decision_tree_impl.hpp

```diff
@@ -56,6 +56,98 @@ DecisionTree<FitnessFunction,
   Train(data, labels, numClasses, minimumLeafSize);
 }
 
+//! Construct and train.
+template<typename FitnessFunction,
+         template<typename> class NumericSplitType,
+         template<typename> class CategoricalSplitType,
+         typename ElemType,
+         bool NoRecursion>
+template<typename MatType>
+DecisionTree<FitnessFunction,
+             NumericSplitType,
+             CategoricalSplitType,
+             ElemType,
+             NoRecursion>::DecisionTree(MatType&& data,
+                                        const data::DatasetInfo& datasetInfo,
+                                        arma::Row<size_t>&& labels,
+                                        const size_t numClasses,
+                                        const size_t minimumLeafSize)
+{
+  // move the data to avoid copy
+  MatType tmpData = std::move(data);
+  arma::Row<size_t> tmpLabels = std::move(labels);
```

Comment:
I realized that this parameter `MatType&&` is a universal reference and can be either an lvalue (`MatType&`) or an rvalue (`MatType&&`). This means that we have the opportunity to do perfect forwarding here. If you changed the labels to also be a universal reference (i.e. `LabelsType&& labels`) then you could do:

```
MatType tmpData(std::forward<MatType>(data));
arma::Row<size_t> tmpLabels(std::forward<LabelsType>(labels));
```

This will copy the data if a reference is given and move the data if an rvalue reference is given, which is exactly the behavior we want.  Then we can remove the other constructor overload that takes a `const MatType&`.  You could do the same with `Train()` too.

I've done some changes just now in another part of the codebase that do a similar thing in 3f7d138, maybe you can look at those as an example.  I regret to say that I had not fully learned about perfect forwarding and `std::forward` in C++ until just today. Learning about it earlier probably could have saved me some pain in the past. Either way I just ordered Scott Meyers' 'Effective Modern C++' so I don't get caught out by stuff like this again. :)